### PR TITLE
[multistage] add broker pruning support to non-partitioned leaf path

### PR DESCRIPTION
**Pull request by dang-stripe**
This addresses part of https://github.com/apache/pinot/issues/18201.

## Overview

* Updates the non-partitioned leaf path by porting [LeafStageToPinotQuery.java](https://github.com/apache/pinot/blob/master/pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/LeafStageToPinotQuery.java) used by physical optimizer to operate on `PlanNode` instead of `RelNode`.
* Updates MSE query response to populate `numSegmentsPrunedByBroker` so we can see how many segments are being pruned (previously was always 0)
* It does not touch the partitioned path or support logical tables yet.

## How to enable pruning

The new broker pruning behavior for plain MSE is off by default. It is gated by:
* **`SET useBrokerPruning=true` query option flag:** This option already existed and defaults to true for the physical optimizer through the [`pinot.broker.multistage.use.broker.pruning`](https://github.com/apache/pinot/blob/b9897956e6c586eb6530102ccf82ec776c18bce3/pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java#L580) broker config flag. This field defaults to false now for plain MSE.
* **`pinot.broker.multistage.logical.planner.use.broker.pruning` broker config**: I decided to use a separate flag from the existing `pinot.broker.multistage.use.broker.pruning` so this could default to off and also be enabled independently of the physical optimizer pruning config.

## Testing

We've replayed prod query traffic internally with this feature off and on and confirm we now see a large reduction servers queried and segments processed. This is with a table that has all segments for each partition colocated to a single server (`numInstancesPerPartition=1`).

```diff
--- without-pruning.json	2026-04-16 10:07:49.505977091 -0700
+++ with-pruning.json	2026-04-16 10:08:07.748201710 -0700
@@ -1,18 +1,18 @@
   "numEntriesScannedInFilter": 149,
   "numEntriesScannedPostFilter": 18340,
-  "numServersQueried": 36,
-  "numServersResponded": 36,
-  "numSegmentsQueried": 44418,
+  "numServersQueried": 7,
+  "numServersResponded": 7,
+  "numSegmentsQueried": 2885,
   "numSegmentsProcessed": 50,
   "numSegmentsMatched": 34,
   "numConsumingSegmentsQueried": 24,
   "numConsumingSegmentsProcessed": 12,
   "numConsumingSegmentsMatched": 0,
-  "minConsumingFreshnessTimeMs": 1776359250213,
-  "numSegmentsPrunedByBroker": 0,
-  "numSegmentsPrunedByServer": 44368,
+  "minConsumingFreshnessTimeMs": 1776359142590,
+  "numSegmentsPrunedByBroker": 41533,
+  "numSegmentsPrunedByServer": 2835,
   "numSegmentsPrunedInvalid": 0,
   "numSegmentsPrunedByLimit": 0,
   "numSegmentsPrunedByValue": 132,
```

cc @Jackie-Jiang @yashmayya @ankitsultana @shauryachats @jadami10

### CODE REVIEW

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java [hunk 1]
```diff
@@ -574,12 +574,21 @@ public static class Broker {
...
+    /**
+     * Whether to use broker pruning by default on the MSE path.
+     * This value can always be overridden by {@link Request.QueryOptionKey#USE_BROKER_PRUNING} query option.
+     * Separated from {@link #CONFIG_OF_USE_BROKER_PRUNING} so the two paths can be rolled out independently.
+     */
+    public static final String CONFIG_OF_MSE_USE_BROKER_PRUNING =
+        "pinot.broker.multistage.mse.use.broker.pruning";
+    public static final boolean DEFAULT_MSE_USE_BROKER_PRUNING = false;
+
     /**
```
[the remark this answers is not in the document]
Comment (dang-stripe) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:592]: fixed with latest commit

File: pinot-query-planner/src/main/java/org/apache/pinot/query/QueryEnvironment.java [hunk 2]
```diff
@@ -797,6 +797,18 @@ default boolean defaultUseBrokerPruning() {
...
+    /**
+     * Whether to use broker pruning by default on the logical planner (non-physical-optimizer) path.
+     *
+     * Separated from {@link #defaultUseBrokerPruning()} so the two planner paths can be rolled out independently.
+     * This default value can be always overridden at query level by the query option
+     * {@link CommonConstants.Broker.Request.QueryOptionKey#USE_BROKER_PRUNING}.
+     */
+    @Value.Default
+    default boolean defaultLogicalPlannerUseBrokerPruning() {
+      return CommonConstants.Broker.DEFAULT_LOGICAL_PLANNER_USE_BROKER_PRUNING;
```
[the remark this answers is not in the document]
Comment (dang-stripe) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/QueryEnvironment.java:809]: fixed with latest commit.

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageBrokerRequestHandler.java [hunk 3]
```diff
@@ -484,6 +485,9 @@ private ImmutableQueryEnvironment.Config getQueryEnvConf(HttpHeaders httpHeaders
     boolean defaultUseBrokerPruning = _config.getProperty(
         CommonConstants.Broker.CONFIG_OF_USE_BROKER_PRUNING,
         CommonConstants.Broker.DEFAULT_USE_BROKER_PRUNING);
+    boolean defaultLogicalPlannerUseBrokerPruning = _config.getProperty(
+        CommonConstants.Broker.CONFIG_OF_LOGICAL_PLANNER_USE_BROKER_PRUNING,
+        CommonConstants.Broker.DEFAULT_LOGICAL_PLANNER_USE_BROKER_PRUNING);
```
[the remark this answers is not in the document]
Comment (dang-stripe) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageBrokerRequestHandler.java:490]: fixed in latest commit

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/physical/DispatchableSubPlan.java [hunk 4]
```diff
@@ -50,14 +50,23 @@ public class DispatchableSubPlan {
   private final Map<Integer, DispatchablePlanFragment> _queryStageMap;
   private final Set<String> _tableNames;
   private final Map<String, Set<String>> _tableToUnavailableSegmentsMap;
+  private final long _numSegmentsPrunedByBroker;
 
   public DispatchableSubPlan(PairList<Integer, String> fields,
       Map<Integer, DispatchablePlanFragment> queryStageMap,
       Set<String> tableNames, Map<String, Set<String>> tableToUnavailableSegmentsMap) {
+    this(fields, queryStageMap, tableNames, tableToUnavailableSegmentsMap, 0);
```
Comment (ankitsultana) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/physical/DispatchableSubPlan.java:58]: allowing multiple flavors of ctors with defaults is quite error prone imo.

could we just keep the single ctor that takes in all the arguments? will also help see whether this value is being passed in correctly or not

File: pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java [hunk 5]
```diff
@@ -624,6 +679,19 @@ private RoutingTable getRoutingTableHelper(String tableNameWithType, long reques
     return _routingManager.getRoutingTable(brokerRequest, requestId);
   }
 
+  @Nullable
+  private RoutingTable getRoutingTableHelper(PinotQuery pinotQuery, long requestId) {
+    return _routingManager.getRoutingTable(CalciteSqlCompiler.convertToBrokerRequest(pinotQuery), requestId);
```
Comment (xiangfu0) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java:684]: This new filtered-routing path drops the original query options before calling the routing manager. The old `getRoutingTableHelper(String, ..., queryOptions)` explicitly copied them onto the synthetic request (and even calls out `sampler`), but `PlanNodeRoutingQueryBuilder` builds a fresh `PinotQuery` with no `queryOptions`, so `CalciteSqlCompiler.convertToBrokerRequest(pinotQuery)` routes as if the query had none. That is a correctness regression for options that affect routing, especially `tableSampler`: once `useBrokerPruning=true` on the logical-planner path, MSE will assign workers from the full segment set instead of the sampled subset and can return unsampled results. Please preserve `context.getPlannerContext().getOptions()` on the routing `PinotQuery` before converting it to a `BrokerRequest`.

Comment (dang-stripe) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java:689]: Good catch - updated to propagate the query options and added a test.

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/physical/DispatchableSubPlan.java (hunk 4 above)
Comment (dang-stripe) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/physical/DispatchableSubPlan.java]: Yep, removed the extra constructor.

**Comment by dang-stripe** [comment]
@ankitsultana @xiangfu0 addressed initial feedback. let me know if there's anything else.

File: pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java [hunk 6]
```diff
@@ -551,6 +554,7 @@ private void assignWorkersToNonPartitionedLeafFragment(DispatchablePlanMetadata
       if (!routingTable.getUnavailableSegments().isEmpty()) {
         metadata.addUnavailableSegments(tableName, routingTable.getUnavailableSegments());
       }
+      context.addNumSegmentsPrunedByBroker(routingTable.getNumPrunedSegments());
```
Comment (gortiz) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java]: `extractRoutingQuery`  is nullable. Shouldn't this produce an NPE whenever this is called with `useBrokerPruning = false`? The same happens with filter

File: pinot-query-planner/src/main/java/org/apache/pinot/query/routing/PlanNodeRoutingQueryBuilder.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    PinotQuery pinotQuery = new PinotQuery();
+    pinotQuery.setDataSource(new DataSource());
+    pinotQuery.getDataSource().setTableName(tableName);
+    pinotQuery.setSelectList(tableScan.getColumns().stream().map(
+        RequestUtils::getIdentifierExpression).collect(Collectors.toList()));
+    return pinotQuery;
+  }
+
+  private static void handleProject(ProjectNode project, PinotQuery pinotQuery) {
+    if (project != null) {
```
Comment (gortiz) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/routing/PlanNodeRoutingQueryBuilder.java]: nit: this check is not needed (right now this method is only called when project is not null). But in case you want to support null projects, we should mark the argument as `@Nullable`

File: pinot-query-planner/src/test/java/org/apache/pinot/query/routing/WorkerManagerTest.java [hunk 8]
```diff
@@ -123,9 +127,215 @@ public void testSingletonWorkerWithEmptyTableAndUseLeafServerEnabled() {
...
+    Map<String, String> tableNameMap = new HashMap<>();
+    tableNameMap.put("testTable_OFFLINE", "testTable_OFFLINE");
+    tableNameMap.put("testTable", "testTable");
+
+    TableCache tableCache = mock(TableCache.class);
+    when(tableCache.getTableNameMap()).thenReturn(tableNameMap);
+    when(tableCache.getActualTableName(anyString())).thenAnswer(inv -> tableNameMap.get(inv.getArgument(0)));
+    when(tableCache.getSchema(anyString())).thenReturn(schema);
+    when(tableCache.getTableConfig("testTable_OFFLINE"))
+        .thenReturn(mock(org.apache.pinot.spi.config.table.TableConfig.class));
```
Comment (gortiz) [inline pinot-query-planner/src/test/java/org/apache/pinot/query/routing/WorkerManagerTest.java]: nit: You are referencing `TableConfig` with the full path several times without importing it. Usually we prefer to use import in these cases

**Review by gortiz — APPROVED** [review]

File: pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java (hunk 6 above)
Comment (dang-stripe) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java]: on line 523, we fall back to the unfiltered routing path when `routingPinotQuery` is null. i added a comment so it's more clear.

File: pinot-query-planner/src/main/java/org/apache/pinot/query/routing/PlanNodeRoutingQueryBuilder.java (hunk 7 above)
Comment (dang-stripe) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/routing/PlanNodeRoutingQueryBuilder.java]: good catch - removed the guards

File: pinot-query-planner/src/test/java/org/apache/pinot/query/routing/WorkerManagerTest.java (hunk 8 above)
Comment (dang-stripe) [reply pinot-query-planner/src/test/java/org/apache/pinot/query/routing/WorkerManagerTest.java]: cleaned it up

**Comment by dang-stripe** [comment]
@gortiz thanks for reviewing! addressed your comments. let me know if there's anything else.

File: pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java (hunk 6 above)
Comment (gortiz) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java]: I would prefer to add the null check just in case something changes later in the code. It should be a very cheap check. Anyway, this is not something blocker

Comment (dang-stripe) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java]: makes sense. i added the null guard. also fixed a test that was missing the pruning flag (adding the null guard surfaced this).

**Comment by dang-stripe** [comment]
@gortiz added the null guard and fixed a test

**Comment by dang-stripe** [comment]
@xiangfu0 @gortiz @ankitsultana @Jackie-Jiang @yashmayya let me know if there's anything else to address. the test failure seems transient (test timed out).

**Comment by dang-stripe** [comment]
friendly ping here

**Comment by xiangfu0** [comment]
Docs update opened in pinot-contrib/pinot-docs#786: https://github.com/pinot-contrib/pinot-docs/pull/786
